src/Lykke.RabbitMq.Azure/DeferredMessagesRepository.cs

```diff
@@ -0,0 +1,199 @@
+        /// </p>
+        /// <p>
+        /// If your publishes about 70 messages per minute and in the average the message is deferred about 1 hour, 
+        /// <see cref="DeferredMessagesPartitionSize.Minute"/> is a good choice. Because a single partition will 
+        /// contains about 70 messages, and there will be about 60 partitions of not overdue messages in the storage
+        /// </p>
+        /// Default value is <see cref="DeferredMessagesPartitionSize.Minute"/>
+        /// </param>
+        public DeferredMessagesRepository(
+            INoSQLTableStorage<DeferredMessageEntity> storage, 
+            DeferredMessagesPartitionSize partitionSize = DeferredMessagesPartitionSize.Minute)
+        {
+            _storage = storage ?? throw new ArgumentNullException(nameof(storage));
+            _partitionSize = partitionSize;
+        }
+
+        /// <inheritdoc cref="IDeferredMessagesRepository.SaveAsync"/>
+        public async Task SaveAsync(byte[] message, DateTime deliverAt)
+        {
+            if (message == null)
+            {
+                throw new ArgumentNullException(nameof(message));
+            }
+
+            var pk = GetPartitionKey(deliverAt);
+            var rk = GetRowKey(deliverAt, Interlocked.Increment(ref _counter));
+
+            await _storage.InsertAsync(new DeferredMessageEntity
+            {
+                PartitionKey = pk,
+                RowKey = rk,
+                Message = message
+            });
+        }
+
+        /// <inheritdoc cref="IDeferredMessagesRepository.GetOverdueMessagesAsync"/>
+        public async Task<IReadOnlyList<DeferredMessageEnvelope>> GetOverdueMessagesAsync(DateTime forTheMoment)
+        {
+            const int resultsLimit = 1000;
+
+            var toPartitionKey = GetPartitionKey(forTheMoment);
+            var partitionFilter = TableQuery.GenerateFilterCondition(
+                nameof(DeferredMessageEntity.PartitionKey), 
+                QueryComparisons.LessThanOrEqual,
+                toPartitionKey);
+            var toRowKey = GetRowKey(forTheMoment.AddTicks(1), 0);
+            var rowFilter = TableQuery.GenerateFilterCondition(
+                nameof(DeferredMessageEntity.RowKey),
+                QueryComparisons.LessThan,
+                toRowKey);
+            var filter = TableQuery.CombineFilters(partitionFilter, TableOperators.And, rowFilter);
+
+            var tableQuery = new TableQuery<DeferredMessageEntity>().Where(filter);
+            var result = new List<DeferredMessageEnvelope>();
+
+            await _storage.ExecuteAsync(tableQuery, messages =>
+                {
+                    var envelopes = messages.Select(message =>
+                        new DeferredMessageEnvelope(
+                            $"{message.PartitionKey}&{message.RowKey}",
+                            message.Message));
+                    result.AddRange(envelopes);
```

Comment:
Is it thread safe? I saw somewhere lock around similar construction

Reply:
Yes, it is thread safe